src/index/Index.cpp

```diff
@@ -121,107 +124,123 @@ template void Index::createFromFile<TurtleMmapParser>(const string& filename);
-      writer << array<Id, 3>{{spoIds[2], langPredId, langTagId}};
-      actualCurrentPartialSize += 2;
-    }
+    std::array<ItemMapManager, NUM_PARALLEL_ITEM_MAPS> itemArray;
 
-    ++i;
-    if (i % 10000000 == 0) {
-      LOG(INFO) << "Lines (from KB-file) processed: " << i << '\n';
+    // that way the different ids won't interfere
+    for (size_t j = 0; j < NUM_PARALLEL_ITEM_MAPS; ++j) {
+      itemArray[j] = ItemMapManager(j * 100 * linesPerPartial);
     }
-    if (i % linesPerPartial == 0) {
-      // wait until sorting the last partial vocabulary has finished
-      // to control the number of threads and the amount of memory used at the
-      // same time. typically sorting is finished before we reach again here so
-      // it is not a bottleneck.
-      if (sortFutures.first.valid()) {
-        sortFutures.first.get();
+
+    const auto itemMapLamdaCreator = [&itemArray](const size_t idx) {
+      return [itPtr = &(itemArray[idx])](LangtagAndTriple&& lt) {
+        OptionalIds res;
+        res[0] = itPtr->assignNextId(lt._triple);
+        if (!lt._langtag.empty()) {
+          auto langTagId = itPtr->assignNextId(
+              ad_utility::convertLangtagToEntityUri(lt._langtag));
+          auto langTaggedPredId =
+              itPtr->assignNextId(ad_utility::convertToLanguageTaggedPredicate(
+                  lt._triple[1], lt._langtag));
+          auto& spoIds = *(res[0]);
+          res[1].emplace(array<Id, 3>{spoIds[0], langTaggedPredId, spoIds[2]});
+          res[2].emplace(array<Id, 3>{
+              spoIds[2], itPtr->getLanguagePredicateId(), langTagId});
+        }
+        return res;
+      };
+    };
+
+    auto itemMapLambdaTuple =
+        ad_tuple_helpers::setupTupleFromCallable<NUM_PARALLEL_ITEM_MAPS>(
+            itemMapLamdaCreator);
+
+    {
+      auto p = ad_pipeline::setupParallelPipeline<1, NUM_PARALLEL_ITEM_MAPS>(
+          _parserBatchSize,
+          [parserPtr = &parser, i = 0ull, linesPerPartial,
+           &parserExhausted]() mutable -> std::optional<Triple> {
+            if (i >= linesPerPartial) {
+              return std::nullopt;
+            }
+            Triple t;
+            if (parserPtr->getLine(t)) {
+              i++;
+              return std::optional(std::move(t));
+            } else {
+              parserExhausted = true;
+              return std::nullopt;
+            }
+          },
+          [this](Triple&& t) {
+            return tripleToInternalRepresentation(std::move(t));
+          },
+          std::move(itemMapLambdaTuple));
+
+      while (auto opt = p.getNextValue()) {
+        i++;
+        for (const auto& innerOpt : opt.value()) {
+          if (innerOpt) {
+            actualCurrentPartialSize++;
+            localWriter << innerOpt.value();
+          }
+        }
+        if (i % 10000000 == 0) {
+          LOG(INFO) << "Lines (from KB-file) processed: " << i << '\n';
+        }
       }
```

Comment:
@joka921 Could you add a few well-placed and concise comments in this code block so that the structure becomes clearer and what the individual lambdas do?

Reply:
I think you mentioned that you could also replace the lambdas with well named functions. I think this might indeed be helpful here